src/app/AttributeAccessInterface.h

```diff
@@ -51,7 +56,31 @@ class AttributeValueEncoder
         {
             return CHIP_NO_ERROR;
         }
-        return DataModel::Encode(*mWriter, TLV::ContextTag(AttributeDataElement::kCsTag_Data), std::forward<Ts>(aArgs)...);
+        return TagBoundTLVEncoder::Encode(std::forward<Ts>(aArgs)...);
+    }
+
+    /**
+     * aCallback is expected to take a const TagBoundTLVEncoder& argument and
+     * Encode() on it as many times as needed to encode all the list elements
+     * one by one.  If any of those Encode() calls returns failure, aCallback
+     * must stop encoding and return failure.  When all items are encoded
+     * aCallback is expected to return success.
+     *
+     * aCallback may not be called.  Consumers must not assume it will be
+     * called.
+     */
+    template <typename ListGenerator>
+    CHIP_ERROR EncodeList(ListGenerator aCallback)
```

Comment:
I think this method should materially be present on the `TagBoundTLVEncoder` class, since the utility of such a method shouldn't be limited to just attributes.

If done that way, then I don't think we actually need the `AttributeValueEncoder` anymore do we? All that remains is the logic for `mTriedEncode`, which honestly, doesn't feel like _enough_ remaining here to justify its existence.

Reply:
I think having this 'contextual writer' object encapsulate not just the tag, but things like the path information (if the cluster object needs to be able to deal with list mutations for example), or as you mentioned, the accessing fabric information (perhaps an ISD-like object), feels like the right direction.

More-over, I feel like the Encode APIs in Encode.h should take this object instead of a `TLVWriter` + tag. This allows consumers of cluster objects to be more insulated about a lot of things they shouldn't have to worry about, unless they need to.

As for the 'mTriedEncode', I think we should drop the 'speculative read' construct in the Engine and bring it back in the future if we need to.

Reply:
As such, I'd rename it `ContextBoundWriter`

Reply:
I'm fine if the above is too ambitious for the now, but we should figure it out soon.